### PR TITLE
Add support for private registry pull secrets

### PR DESCRIPTION
**Pull request by lkolacek**

### CODE REVIEW

File: iib/workers/tasks/build.py
Comment (lkolacek) [inline iib/workers/tasks/build.py:957]: @lcarva do we want to replace it for https://github.com/release-engineering/iib/blob/master/iib/workers/config.py#L15 ??

File: iib/web/static/api_v1.yaml
Comment (lkolacek) [inline iib/web/static/api_v1.yaml:755]: @lcarva which iib operations should have registry_auths added?  Add operator bundle, remove operator, anything else?

File: iib/workers/tasks/build.py
Comment (lkolacek) [inline iib/workers/tasks/build.py:1026]: @lcarva looks like with dockerconfig added I was able to inspect image in registry.redhat.io in local dev environment  http://pastebin.test.redhat.com/948693

File: iib/web/static/api_v1.yaml
Comment (lkolacek) [inline iib/web/static/api_v1.yaml:814]: @lcarva I used string type for this parameter. On web it's displayed like this: 
```
{
"add_arches": [
"amd64",
"s390x"
],
"binary_image": "quay.io/operator-framework/upstream-registry-builder:v1.5.9",
"bundles": [
"quay.io/user/openshift-bundle-example-container:v1"
],
"cnr_token": "token",
"deprecation_list": [
"quay.io/user/openshift-bundle-example-container@sha256:29e4531e4625653"
],
"force_backport": false,
"from_index": "quay.io/iib-stage/iib:4",
"organization": "openshift-operators",
"overwrite_from_index": false,
"overwrite_from_index_token": "token",
"registry_auths": "{\"auths\": {\"registry.example.io\": {\"auth\": \"my-secret-token-base-64-encoded\"}}}",
"distribution_scope": "prod"
}
```
If correct registry and secret is provided in this format ^ . It works well. Let me know if yo uhave better idea how to deal with passing json.

**Comment by lkolacek** [comment]
PTAL @yashvardhannanavati

File: iib/web/models.py
Comment (lcarva) [inline iib/web/models.py:928]: Is this needed? You're already specifying it in `RequestIndexImageMixin._from_json`. I think we only need to specify it in one place. If this is only to be used for `add` requests, and not `rm`, then I think it should only be listed here and not in `RequestIndexImageMixin._from_json`.

File: iib/web/static/api_v1.yaml
Comment (lcarva) [reply iib/web/static/api_v1.yaml:755]: That's a good question. It's needed in at least `regenerate-bundle`. Let's follow up with the CVP on this.

File: iib/web/static/api_v1.yaml
Comment (lcarva) [reply iib/web/static/api_v1.yaml:814]: I think it's odd to inline a json inside another json. Can we use a json object instead of string?

File: iib/workers/tasks/build.py
Comment (lcarva) [reply iib/workers/tasks/build.py:957]: Let's use the existing `set_registry_token` context manager: https://github.com/release-engineering/iib/blob/c8b668b95d41b9d4807a279a1b1154f2296930fe/iib/workers/tasks/utils.py

File: iib/web/static/api_v1.yaml
Comment (lcarva) [reply iib/web/static/api_v1.yaml:755]: Confirmed with them. Let's just add it to `regenerate-bundle`.

File: iib/workers/tasks/build.py
Comment (lkolacek) [reply iib/workers/tasks/build.py]: I added adjustment to `set_registry_token` then, hope this is close to what you had in mind.

Comment (lkolacek) [reply iib/workers/tasks/build.py]: If it looks usable it's probably worth to rename that method.

File: iib/web/models.py
Comment (lcarva) [inline iib/web/models.py:1096]: We should validate that if `registry_auths` is used, it conforms to the expected format.

File: iib/workers/tasks/build.py
Comment (lcarva) [inline iib/workers/tasks/build.py:1248]: `dict` not a `str`, right?

File: iib/workers/tasks/utils.py
Comment (lcarva) [inline iib/workers/tasks/utils.py:243]: Looking at the different conditionals in this function now, would it make sense to have a new `set_registry_auths` function that gets used by `set_registry_token`?

```python
def set_registry_token(token, container_image):
    if not token:
        ...
    if not container_image:
        ...
    registry = ImageName.parse(container_image).registry
    encoded_token = base64.b64encode(token.encode('utf-8')).decode('utf-8')
    registry_auths = {'auths': {registry: encoded_token}}
    with set_registry_auths(registry_auths):
      yield
    return


def set_registry_auths(registry_auths):
    if not registry_auths:
        ...
    docker_config_path = os.path.join(os.path.expanduser('~'), '.docker', 'config.json')
    try:                                                                                            
        log.debug('Removing the Docker config symlink at %s', docker_config_path)                      
        try:                                                                                        
            os.remove(docker_config_path)                                                           
        except FileNotFoundError:                                                                   
            log.debug('The Docker config symlink at %s does not exist', docker_config_path)         
                                                                                                    
        conf = get_worker_config()                                                                  
        if os.path.exists(conf.iib_docker_config_template):                                         
            with open(conf.iib_docker_config_template, 'r') as f:                                   
                docker_config = json.load(f)                                                        
        else:                                                                                       
            docker_config = {}                                                                      
                                                                
        registries = list(registry_auths.get('auths', {}).keys())
        log.debug('Setting the override token for the registries %s in the Docker config', registries)  

        docker_config.setdefault('auths', {})
        docker_config['auths'].update(registry_auths.get('auths', {}))                              
        with open(docker_config_path, 'w') as f:                                                    
            json.dump(docker_config, f)                                                             
                                                                                                    
        yield                                                                                       
    finally:                                                                                        
        reset_docker_config()
```

Beware that I have not tested the code above.

File: iib/web/models.py
Comment (lkolacek) [reply iib/web/models.py]: I added check if it's a dict, not sure if we need a more sophisticated check here.

Comment (lkolacek) [reply iib/web/models.py]: Do you want check if there is 'auths' on top level and then registries?

File: iib/workers/tasks/build.py
Comment (lkolacek) [reply iib/workers/tasks/build.py]: Yes, good catch

File: iib/workers/tasks/utils.py
Comment (lkolacek) [reply iib/workers/tasks/utils.py]: Nice! It makes perfect sense

File: iib/web/models.py
Comment (lcarva) [reply iib/web/models.py]: The `.dockerconfig.json` format allows for things other than auth information to be set. We should prevent those from being used. I think the most straightforward way is to raise a validation error if those are set. So, I propose you check the following:

1. `registry_auths` is of type dict
2. `regsitry_auths` contain a single key, `auths`
3. `regsitry_auths.auths` is a dict, it could be empty IMO
4. each _value_ in `auths` is a dict and must have one and only key named `auth`. The value of `auth` should be a string.

This will look verbose, but we don't really have another option. At some point we should revisit using a json schema to do this sort of thing.

File: iib/workers/tasks/utils.py
Comment (lcarva) [inline iib/workers/tasks/utils.py:249]: Is there a particular reason why this empty line, and 2 above, were removed?

File: iib/web/static/api_v1.yaml
Comment (lcarva) [inline iib/web/static/api_v1.yaml:844]: Let's state here that non-auth information in the dockerconfig.json is not allowed.

**Review by lcarva — COMMENTED** [review]
This is looking really good!

**Comment by lcarva** [comment]
@lkolacek, checkout the output here: https://travis-ci.com/github/release-engineering/iib/jobs/495018377
Run `tox` locally to reproduce the linting errors.

File: iib/workers/tasks/utils.py
Comment (lkolacek) [reply iib/workers/tasks/utils.py:249]: nope were removed accidently

File: iib/web/static/api_v1.yaml
Comment (lkolacek) [reply iib/web/static/api_v1.yaml]: Changed

File: iib/web/models.py
Comment (lkolacek) [reply iib/web/models.py]: I added validation check. Looks terrible and verbose. Please take a look if it looks usable @lcarva . I suppose unit test is required. :)

File: iib/web/models.py
Comment (lcarva) [inline iib/web/models.py:1426]: Let's use the variable `auths` in the error message too.

File: iib/web/models.py
Comment (lcarva) [inline iib/web/models.py:1428]: Let's use the variable `auths` in the error message too.

File: iib/web/models.py
Comment (lcarva) [reply iib/web/models.py]: It doesn't look that bad :)

File: tests/test_workers/test_tasks/test_utils.py
Comment (lcarva) [inline tests/test_workers/test_tasks/test_utils.py:117]: Let's add a test where we validate that the auth info from `registry_auths` overwrites whatever is in the config.

Comment (lkolacek) [reply tests/test_workers/test_tasks/test_utils.py:117]: I added 1 extra registry which should be overwritten with provided registry_auths. Does it look sufficient? https://github.com/release-engineering/iib/pull/227/files#diff-ad30f2cc32db90c116232cb76cc707702b22936e22f90e0fd5f5d215aa984ec9R134

File: iib/web/models.py
Comment (lkolacek) [reply iib/web/models.py]: Variable added yo err msg.

File: iib/web/models.py
Comment (lkolacek) [reply iib/web/models.py]: Variable added yo err msg.

**Review by lcarva — APPROVED** [review]

File: iib/web/models.py
Comment (yashvardhannanavati) [inline iib/web/models.py:1430]: How about changing this to `{reg} in registry_auths has auth value in an incorrect format. It should be a base64 encoded token`?

**Review by yashvardhannanavati — APPROVED** [review]

File: iib/web/models.py
Comment (lcarva) [reply iib/web/models.py:1430]: It's more complex than that. It's well described in the API docs. How about something like `{reg} in registry_auths has auth value in an incorrect format. See the API docs for details on the expected format` ?

Comment (lkolacek) [reply iib/web/models.py]: I added second option `See the API docs for details on the` ..

**Comment by lkolacek** [comment]
@yashvardhannanavati  @lcarva Once your happy with changes, please merge this PR. I don't have permissions.

**Comment by yashvardhannanavati** [comment]
I'm so sorry. My github UI was outdated and I accidentally hit the merge button.

**Comment by yashvardhannanavati** [comment]
@lkolacek Could you please address Paya's comments in a separate PR? Or would you like me to revert the change? 
@pbortlov If your comments have major changes, I can revert the commit

**Comment by lkolacek** [comment]
@yashvardhannanavati  np I'll address comments in separate PR

File: iib/workers/tasks/build.py
Comment (lkolacek) [reply iib/workers/tasks/build.py:1258]: set_registry_auths has contextmanager decorator and it applies changes to dockerconfig.json only when registry_auths were provided. See https://github.com/release-engineering/iib/pull/227/files#diff-1bece9798c37acfe1ca705657d12580325f65a8edd965d6c3d004c7fa3237b40R290  and https://docs.python.org/3/library/contextlib.html

**Pull request merged**
